### PR TITLE
Replace Google Charts with Chart.js

### PR DESCRIPTION
**Pull request by fanthos**
I am opening this PR for discuss some detailed information about how to implement the charts with working code.
Currently, I am using Billboard.js and D3, they costs about 380kb when build as Polymer fragments and 120kb gzipped. I am not sure if there is other way to reduce the size(compared with 300kb+ gzipped Google Charts).
The timeline chart is modified from https://github.com/jjagielka/d3-gtimeline .

### CODE REVIEW

**Comment by fanthos** [comment]
Related issue: #349

**Comment by andrey-git** [comment]
Is all of this JS HA-specific or is it a fork of another code?

**Comment by fanthos** [comment]
https://github.com/fanthos/home-assistant-polymer/blob/baa6b79ccaf3a5a4bcffc623f2dc1b54626a415f/src/components/entity/ha-chart-base.html#L48-L437
These code forked from [d3-gtimeline](https://github.com/jjagielka/d3-gtimeline) under BSD and modified for HA. Other code are HA specific.
Bower dependency Billboardjs licensed under MIT.

**Comment by balloob** [comment]
If you fork a library, please do not include it in our codebase but instead publish it to NPM instead.

File: src/components/entity/ha-chart-base.html
Comment (balloob) [inline src/components/entity/ha-chart-base.html]: These should not be inside the dom-module.

File: src/components/entity/ha-chart-base.html
Comment (balloob) [inline src/components/entity/ha-chart-base.html]: Why was it modified and rewritten? Would it be possible to contribute back to the original library the functionality that we need?

File: src/components/entity/ha-chart-line.html
Comment (balloob) [inline src/components/entity/ha-chart-line.html]: Please use new Polymer 2 class syntax

File: src/components/entity/ha-chart-line.html
Comment (balloob) [inline src/components/entity/ha-chart-line.html]: I thought that native was being used on supported platforms. Why toggle this? We shouldn't do this.

File: src/components/entity/ha-chart-line.html
Comment (balloob) [inline src/components/entity/ha-chart-line.html]: Stale comment

File: src/components/entity/ha-chart-timeline.html
Comment (balloob) [inline src/components/entity/ha-chart-timeline.html]: Please use Polymer 2 class syntax.

File: src/components/entity/ha-chart-line.html
Comment (fanthos) [reply src/components/entity/ha-chart-line.html]: Will merge with history-chart-line later.

File: src/components/entity/ha-chart-base.html
Comment (fanthos) [reply src/components/entity/ha-chart-base.html]: Some feature can push back to the original one, but the data structure is changed. I moved the chart code to standalone Github repo.

**Comment by fanthos** [comment]
Right now bower_components/billboard.js wouldn't build with script/build_frontend because babel is trying to compile billboard.js. Currently I am using prebuilt version located in `billboard.js/dist`, so I think that I don't have to rebuild the dependency with babel. Not sure what I should do with it.
```
[17:17:42] Using gulpfile /mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/gulpfile.js
[17:17:42] Starting 'default'...
[17:17:42] Starting 'clean'...
[17:17:42] Finished 'default' after 4.9 ms
[17:17:42] Finished 'clean' after 24 ms
[17:17:42] Starting 'run_rollup'...
[17:17:46] Finished 'run_rollup' after 4.12 s
[17:17:46] Starting 'ru_all'...
[17:17:46] Finished 'ru_all' after 1.13 ms
[17:17:46] Starting 'build'...

events.js:182
      throw er; // Unhandled 'error' event
      ^
ReferenceError: Unknown plugin "add-module-exports" specified in "/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/bower_components/billboard.js/.babelrc" at 0, attempted to resolve relative to "/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/bower_components/billboard.js"
    at /mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/babel-core/lib/transformation/file/options/option-manager.js:180:17
    at Array.map (native)
    at Function.normalisePlugins (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/babel-core/lib/transformation/file/options/option-manager.js:158:20)
    at OptionManager.mergeOptions (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/babel-core/lib/transformation/file/options/option-manager.js:234:36)
    at OptionManager.init (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/babel-core/lib/transformation/file/options/option-manager.js:368:12)
    at File.initOptions (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/babel-core/lib/transformation/file/index.js:212:65)
    at new File (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/babel-core/lib/transformation/file/index.js:135:24)
    at Pipeline.transform (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/babel-core/lib/transformation/pipeline.js:46:16)
    at DestroyableTransform._transform (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/gulp-babel/index.js:36:22)
    at DestroyableTransform.Transform._read (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/readable-stream/lib/_stream_transform.js:182:10)
    at DestroyableTransform.Transform._write (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/readable-stream/lib/_stream_transform.js:170:83)
    at doWrite (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/readable-stream/lib/_stream_writable.js:406:64)
    at writeOrBuffer (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/readable-stream/lib/_stream_writable.js:395:5)
```

**Comment by balloob** [comment]
Can be fixed by disabling the use of `babelrc` in the build script. On this line: https://github.com/home-assistant/home-assistant-polymer/blob/master/gulp/common/transform.js#L14 add `babelrc: false,`

**Comment by fanthos** [comment]
Not sure if this change will affect other denpendency, will check tomorrow.

**Comment by fanthos** [comment]
Seems works as excepted

**Comment by fanthos** [comment]
I think the code can consider finished.
The houndci keep failing but I don't know the reason.

File: src/components/entity/ha-chart-base.html
Comment (balloob) [inline src/components/entity/ha-chart-base.html]: this needs to import polymer-element.

File: src/components/entity/ha-chart-base.html
Comment (balloob) [inline src/components/entity/ha-chart-base.html]: Note to self to not merge this until I have reworked this to be loaded on demand.

File: src/components/entity/ha-chart-base.html
Comment (balloob) [inline src/components/entity/ha-chart-base.html]: we should name classes after the custom element name, so in this case HaChartBase

File: src/components/entity/ha-chart-base.html
Comment (balloob) [inline src/components/entity/ha-chart-base.html]: This makes no sense, if it should be only included once it should be only loaded once.c

File: src/components/entity/ha-chart-base.html
Comment (balloob) [inline src/components/entity/ha-chart-base.html]: stale comment

File: src/components/entity/ha-chart-base.html
Comment (balloob) [inline src/components/entity/ha-chart-base.html]: Please merge with previous statement.

File: src/components/entity/ha-chart-base.html
Comment (balloob) [inline src/components/entity/ha-chart-base.html]: stale comment

File: src/components/entity/ha-chart-base.html
Comment (balloob) [inline src/components/entity/ha-chart-base.html]: This will never be set because you initialize `chartData.data` to not have this attribute.

File: src/components/entity/ha-chart-base.html
Comment (balloob) [inline src/components/entity/ha-chart-base.html]: Any reason this cannot be included in the object definition above?

File: src/components/entity/ha-chart-base.html
Comment (balloob) [inline src/components/entity/ha-chart-base.html]: This makes no sense, why not use `data` ? 

Also, we should never call anything new, it is bound to be not the new thing in the future.

File: src/components/state-history-chart-line.html
Comment (balloob) [inline src/components/state-history-chart-line.html]: selected-element is a read-only property. This should not work ?

File: src/components/state-history-chart-line.html
Comment (balloob) [inline src/components/state-history-chart-line.html]: stale comment.

File: src/components/state-history-chart-line.html
Comment (balloob) [inline src/components/state-history-chart-line.html]: `dataTables` is an array, no reason to use `Array.prototype`

File: src/components/state-history-chart-line.html
Comment (balloob) [inline src/components/state-history-chart-line.html]: selectedDataPoint is not being used nor defined

File: src/components/state-history-chart-timeline.html
Comment (balloob) [inline src/components/state-history-chart-timeline.html]: selectedDataPoint not being used/defined.

File: src/components/state-history-chart-timeline.html
Comment (balloob) [inline src/components/state-history-chart-timeline.html]: Can merge this if with the one above.

Comment (balloob) [reply src/components/state-history-chart-timeline.html]: Actually, this check is now wrong because you removed the code that made the div all empty. Now if the new data is empty, it will not remove the old one and thus show stale data

**Comment by balloob** [comment]
This PR has conflicts now 😞

**Comment by fanthos** [comment]
I will look into this later

**Comment by balloob** [comment]
This PR seems to have gone stale. Closing it. You can reopen it when you're ready to finish it.

**Comment by fanthos** [comment]
I had more work on this PR, it seems work now.
Should I re-open the PR or open a new PR?

**Comment by fanthos** [comment]
@balloob 
I am not able to reopen this PR.
Should I open a new PR or someone can reopen this one?

**Comment by homeassistant** [comment]
Hello @fanthos,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`ee2b9c5122efbf3278a4cad5f18cbe9b37766fe0`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant-polymer&#x2F;pull&#x2F;429&#x2F;commits&#x2F;ee2b9c5122efbf3278a4cad5f18cbe9b37766fe0) - This commit has something that looks like an email address (dllcadugmiet@web.de). Maybe try linking that to GitHub?.
- [`017815141150f5749a0b6faa6cf99c0fc08bcecf`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant-polymer&#x2F;pull&#x2F;429&#x2F;commits&#x2F;017815141150f5749a0b6faa6cf99c0fc08bcecf) - This commit has something that looks like an email address (dllcadugmiet@web.de). Maybe try linking that to GitHub?.
- [`7b61826134edb0e1a0e93d0346009478eb7f6ba5`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant-polymer&#x2F;pull&#x2F;429&#x2F;commits&#x2F;7b61826134edb0e1a0e93d0346009478eb7f6ba5) - This commit has something that looks like an email address (dllcadugmiet@web.de). Maybe try linking that to GitHub?.
- [`7d86f30a6b30eb2174a56d03cd5ea4bf500ab03d`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant-polymer&#x2F;pull&#x2F;429&#x2F;commits&#x2F;7d86f30a6b30eb2174a56d03cd5ea4bf500ab03d) - This commit has something that looks like an email address (dllcadugmiet@web.de). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

File: src/components/entity/ha-chart-base.html
Comment (andrey-git) [inline src/components/entity/ha-chart-base.html]: I don't think `readOnly: true` has any effect.

File: src/components/entity/ha-chart-base.html
Comment (andrey-git) [inline src/components/entity/ha-chart-base.html]: Make a single observer that will observe all the variables instead. Otherwise if both `data` and `connected` change together the observer would be called twice.

File: src/components/entity/ha-chart-base.html
Comment (andrey-git) [reply src/components/entity/ha-chart-base.html]: Also, is this supposed to be a boolean?

File: src/components/entity/ha-chart-base.html
Comment (andrey-git) [inline src/components/entity/ha-chart-base.html]: call `super.connectedCallback()`

File: src/components/entity/ha-chart-base.html
Comment (andrey-git) [inline src/components/entity/ha-chart-base.html]: I don't think Polymer2 has `this.async` ...

**Review by andrey-git — COMMENTED** [review]
Is this ready for review?

**Comment by fanthos** [comment]
No idea with how to fix this issue:
image

**Comment by fanthos** [comment]
I think this PR is ready for review.

**Review by michaelarnauts — COMMENTED** [review]
There is still some commented out code that should be cleaned up. Thanks for the effort!

**Review by andrey-git — COMMENTED** [review]
chartsjs

In timeline mode the name is cut.
Is it possible to put the state (For example "on") in the timeline chart like in google charts?

File: src/components/entity/ha-chart-base.html
Comment (andrey-git) [reply src/components/entity/ha-chart-base.html]: edit: Actually just add it to bower.json, as timeline doesn't work with minimized momentsjs

**Comment by armills** [comment]
Also, how large is the included moment.js? I know we've tried to avoid it in the past due to the file size.

File: src/components/entity/ha-chart-base.html
Comment (andrey-git) [reply src/components/entity/ha-chart-base.html]: Actually those should be loaded on demand. Putting them in script.src will pull them into frontend.html

**Comment by andrey-git** [comment]
Moment is 130k, 52k minimized.

In any case those should be loaded on demand and not bundled into frontend.html

File: src/components/state-history-chart-line.html
Comment (andrey-git) [inline src/components/state-history-chart-line.html]: I think there is no need to do string manipulations - just pass fill and step as params.

File: src/components/state-history-chart-line.html
Comment (andrey-git) [inline src/components/state-history-chart-line.html]: I think we should add the fill for both 'heat
 and 'cool' as separate columns.

However maybe we should check if at least one datapoint exists - otherwise don't add the column.

**Review by andrey-git — COMMENTED** [review]
Is it possible to show legend, like in google charts?
I think it is useful for charts with a few lines, like climate

File: src/components/entity/ha-chart-base.html
Comment (fanthos) [reply src/components/entity/ha-chart-base.html]: I think the whole ha-chart-base should load on demand because currently only Chart.js depends on moment.js.

Comment (fanthos) [reply src/components/entity/ha-chart-base.html]: I didn't test timeline chart with the bundled version.

File: bower.json
Comment (fanthos) [reply bower.json]: I think the advantage is easier to use it on other projects. The disadvantage is if there's some bug, people may need to fix it in two repo.

**Comment by fanthos** [comment]
Line chart has legend support, so it can be enabled.

**Comment by balloob** [comment]
Yeah we should make sure we load the chart JavaScript and moment.js on demand, they are big. I don't mind adding the code to load it on demand after this PR is merged.

**Comment by andrey-git** [comment]
I think on-demand load is a must. If we merge this PR without on-demand load we will hurt load performance for everyone for sake of better graphs.

**Comment by fanthos** [comment]
fixed

**Comment by fanthos** [comment]
I think this PR can consider ready. Please let me know if I missed something.

File: src/components/state-history-chart-timeline.html
Comment (balloob) [inline src/components/state-history-chart-timeline.html]: This should still be polymer-element (that's what the class extends from)

File: bower.json
Comment (balloob) [inline bower.json]: You need to add moment too for Chart.js to work.

**Comment by balloob** [comment]
image

Font inside the timeline is using a serif font. Can we change that to Roboto (or at least a sans-serif font)?

File: src/components/entity/ha-chart-base.html
Comment (balloob) [inline src/components/entity/ha-chart-base.html]: Instead of executing it inside an IIFE, you can just wrap it inside a block:

```js
{
  Chart.Interaction.modes.neareach = …
}

```

File: src/components/entity/ha-chart-base.html
Comment (balloob) [inline src/components/entity/ha-chart-base.html]: We're using es2015 so we should use const and let.

File: src/components/entity/ha-chart-base.html
Comment (balloob) [inline src/components/entity/ha-chart-base.html]: Use a lambda: `return elements.filter(n => n !== undefined)`

File: src/components/entity/ha-chart-base.html
Comment (balloob) [inline src/components/entity/ha-chart-base.html]: Use a guard clause to keep the code readable

```js
if (!this.isTimeline) {
  this._chart.resize();
  return;
}

if (data.datasets.length === 0) return;

const axis = …

```

File: src/components/state-history-chart-line.html
Comment (balloob) [inline src/components/state-history-chart-line.html]: Only works inside functions. You can remove it.

File: src/components/state-history-chart-line.html
Comment (balloob) [inline src/components/state-history-chart-line.html]: `const`

File: src/components/state-history-chart-line.html
Comment (balloob) [inline src/components/state-history-chart-line.html]: stale comment?

**Comment by balloob** [comment]
Alright, sorry a code review took so long. Looking at graphing code is not my favorite 😉 

Gist of my comments are:
 - Get rid of the usage of `var`, use `const` and `let` instead.
 - Use lambda functions (`=>`) as much as possible
 - Move all chart/moment imports and the `nearest` mode to a separate HTML file that is then imported. I will make sure to extract that HTML import and have it be loaded on demand.

**Comment by balloob** [comment]
I'll make sure to code review it again as soon as possible once the fixes have been done.

I want to aim to get this in before release 0.64. That is going to be released in 2.5 weeks.

**Comment by fanthos** [comment]
Thanks.
I moved the scripts to a separated file but don't know which folder should I put it.

**Comment by balloob** [comment]
You can add it to `src/resources`.

**Comment by balloob** [comment]
Note that the tests also have detected an error of an undefined property.

**Review by balloob — APPROVED** [review]
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉

**Comment by balloob** [comment]
I will update it tomorrow to do the code splitting. Thanks for your patience and hard work!

**Comment by andrey-git** [comment]
Was on-demand loading implemented?

**Comment by balloob** [comment]
I will add that tomorrow

**Comment by balloob** [comment]
Added 👍

**Comment by balloob** [comment]
Just realized this PR has been 6 months in the making.

**Comment by balloob** [comment]
Verified this PR works with both ES5 and latest in release mode. Niceeeee.

File: script/build_frontend
Comment (andrey-git) [inline script/build_frontend]: Don't we need to fingerprint it?

Comment (balloob) [reply script/build_frontend]: I can't wait to Polymer 3 and native code splitting support via rollup/Webpack. No more hacky scripts on our part

**Comment by andrey-git** [comment]
Why is the legend generated manually and not by chartsjs?

**Comment by fanthos** [comment]
The chartjs one draws legend on canvas and hard to calculate chart height.

**Comment by deel77** [comment]
@fanthos I just want to ask, if it was some specific requirement to use "stepped" graphs (http://www.chartjs.org/samples/latest/charts/line/stepped.html)? Is there a way to change it?

**Comment by fanthos** [comment]
@deel77 I am using the "stepped" to replace "interpolate" on Google charts and try to keep the lines looks similar as before.
According the code, the line will draw stepped if the data belongs a "sensor" domain.

**Comment by balloob** [comment]
The problem is that we don't know if the value of a sensor can be interpolated.

In Home Assistant, a state will only change if the value is different. The exact same state will be treated as a new state if the entity property `force_update` is set to True. This allows us to interpolate the data from this entity. However, we don't record that this entity has force_update set to True, so we don't know

**Pull request merged**
